src/libgit2/transports/smart_protocol.c

```diff
@@ -433,6 +433,7 @@ int git_smart__negotiate_fetch(
 
 		if ((error = git_smart__negotiation_step(&t->parent, data.ptr, data.size)) < 0)
 			goto on_error;
+		git_str_clear(&data);
```

Comment:
You are indeed - I think that you need to differentiate the clearing of data based on whether we're in RPC mode (HTTP) or not (SSH).

```suggestion

		if (!t->rpc)
			git_str_clear(&data);
```

?